test/uf_test.cc

```diff
@@ -118,70 +118,153 @@ TEST(ZombieUFTest, ReplayChangeCost) {
+    EXPECT_FALSE(z3.evicted());
+  }
 
-    Zombie<int> z1 = bindZombie([&]() {
-      t.meter.fast_forward(10s);
-      return Zombie<int>(1);
-    });
-    Zombie<int> z2 = bindZombie([&](int x1) {
-      t.meter.fast_forward(11s);
-      return Zombie<int>(x1 + 1);
-    }, z1);
+  {
+    using namespace UFCost;
+    EXAMPLE_CODE
+    EXPECT_TRUE (z3.evicted());
+    EXPECT_FALSE(z2.evicted());
+  }
+}
 
 
-    Zombie<int> z3 = bindZombie([&]() {
-      t.meter.fast_forward(20s);
-      return Zombie<int>(4);
-    });
 
-    t.meter.fast_forward(1s);
 
-    t.reaper.murder();
-    EXPECT_TRUE (z1.evicted());
-    EXPECT_FALSE(z2.evicted() || z3.evicted());
+// access a list of zombies with linear dependency, first from beginning to end,
+// then from end to beginning.
+// - total_size: total number of zombies
+// - memory_limit: number of zombies alive
+// - return value: the list of not evicted zombies after the test
+template<typename test_id>
+ns LinearDependencyForwardBackwardTest(unsigned int total_size, unsigned int memory_limit) {
+  assert(total_size > 1);
+  assert(0 < memory_limit && memory_limit <= total_size);
 
-    t.reaper.murder();
-    EXPECT_TRUE (z2.evicted());
-    EXPECT_FALSE(z3.evicted());
-  }
+  using namespace UF;
+  auto& t = Trailokya::get_trailokya();
 
-  {
-    using namespace UF;
-    auto& t = Trailokya::get_trailokya();
+  using Resource = Resource<test_id>;
 
-    Zombie<int> z1 = bindZombie([&]() {
-      t.meter.fast_forward(10s);
-      return Zombie<int>(1);
-    });
-    Zombie<int> z2 = bindZombie([&](int x1) {
-      t.meter.fast_forward(11s);
-      return Zombie<int>(x1 + 1);
-    }, z1);
+  std::vector<Zombie<Resource>> zs;
 
+  auto allocate_memory = [&]() {
+    while (Resource::count >= memory_limit)
+      t.reaper.murder();
+  };
 
-    Zombie<int> z3 = bindZombie([&]() {
-      t.meter.fast_forward(20s);
-      return Zombie<int>(4);
-    });
 
+  ns t0 = t.meter.time();
+  zs.push_back(bindZombie([&]() {
+    allocate_memory();
+    t.meter.fast_forward(100s);
+    return Zombie<Resource>(0);
+  }));
+
+  // forward
+  for (int i = 1; i < total_size; ++i) {
+    zs.push_back(bindZombie([&](const Resource& x) {
+      allocate_memory();
+      t.meter.fast_forward(100s);
+      return Zombie<Resource>(x.value + 1);
+    }, zs[i - 1]));
+  }
 
-    t.meter.fast_forward(1s);
 
-    t.reaper.murder();
-    EXPECT_TRUE (z1.evicted());
-    EXPECT_FALSE(z2.evicted() || z3.evicted());
+  // backward
+  for (int i = total_size - 1; i >= 0; --i) {
+    t.meter.fast_forward(10s);
+    Zombie<Resource> z = zs[i];
+    EXPECT_EQ(z.shared_ptr()->get_ref().value, i);
+    EXPECT_FALSE(z.evicted());
+  }
+  ns t1 = t.meter.time();
 
-    std::cout << t.book.get_aff(0) << std::endl;
-    std::cout << t.book.get_aff(1) << std::endl;
-    t.meter.fast_forward(1s);
+  while (t.reaper.have_soul())
     t.reaper.murder();
-    EXPECT_TRUE (z3.evicted());
-    EXPECT_FALSE(z2.evicted());
+
+  return t1 - t0;
```

Comment:
can we, instead of measuring time, use a static var to track amount of work done? everytime resource increase it advance by 1. this will remove some noise in the test.